src/Compilers/Core/Portable/Diagnostic/DiagnosticInfo.cs

```diff
@@ -165,7 +165,21 @@ internal DiagnosticInfo(CommonMessageProvider messageProvider, bool isWarningAsE
         }
 
         // Create a copy of this instance with a explicit overridden severity
-        internal virtual DiagnosticInfo GetInstanceWithSeverity(DiagnosticSeverity severity)
+        internal DiagnosticInfo GetInstanceWithSeverity(DiagnosticSeverity severity)
+        {
+            if (Severity != severity)
+            {
+                var result = GetInstanceWithSeverityCore(severity);
+                // If this assert fails it means a subtype of DiagnosticInfo failed to override GetInstanceWithSeverityCore
+                Debug.Assert(this.GetType() == result.GetType());
```

Comment:
This doesn't solve the problem when the missing override is in `abstract` type in the middle of the inheritance hierarchy right?

Reply:
I think we don't mind if an abstract type is missing the override. We just care if a type that was actually created is missing it--if some subtype went to the trouble of declaring it, I guess we just have to hope it didn't drop information from its base type :P